File: src/main/java/gameboyemu/system/WRAMimpl.java
package gameboyemu.system;


/**
 User: par
 Date: 2012-02-04
 */
public class WRAMimpl implements WRAM {


    private int[] mem = new int[4096*2];

    public void writeByte(int address, int b) {
        mem[address - 0xc000] = b;
    }

    public int readByte(int address) {
        return mem[address - 0xc000];
    }

    public void writeWord(int address, int w) {
        int lowByte = w & 0xff;
        int hiByte = (w & 0xff00) >> 8;
        writeByte(address, lowByte);
        writeByte(address + 1, hiByte);
    }

    public int readWord(int address) {
        int lowByte = readByte(address);
        int hiByte = readByte(address + 1);
        return (hiByte << 8)  | lowByte;
    }
}

File: shared/src/main/scala/gameboyemu/system/WRAMimpl.scala
package gameboyemu.system

/**
  * User: par
  * Date: 2012-02-04
  */
class WRAMimpl extends WRAM {
  private val mem = new Array[Int](4096 * 2)

  override def writeByte(address: Int, b: Int): Unit = mem(address - 0xc000) = b

  override def readByte(address: Int): Int = mem(address - 0xc000)

  def writeWord(address: Int, w: Int): Unit = {
    val lowByte = w & 0xff
    val hiByte = (w & 0xff00) >> 8
    writeByte(address, lowByte)
    writeByte(address + 1, hiByte)
  }

  def readWord(address: Int): Int = {
    val lowByte = readByte(address)
    val hiByte = readByte(address + 1)
    (hiByte << 8) | lowByte
  }
}